### PR TITLE
Standalone implementation for empty file collection snapshot

### PR DESCRIPTION
**Pull request by lptr**

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    private static final HashCode SIGNATURE = Hashing.md5().hashString(FileCollectionSnapshot.class.getName() + "$EMPTY");
+
+    @Override
+    public boolean isEmpty() {
+        return true;
+    }
+
+    @Override
+    public Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, String title, boolean includeAdded) {
+        return Iterators.emptyIterator();
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java]: ❌ This is wrong. If the oldSnapshot is non-empty, then all the files in the oldSnapshot should be reported as removed.

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=10892115

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java (hunk 1 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java]: Ah, you're right...

**Comment by lptr** [comment]
New build: https://builds.gradle.org/viewQueued.html?itemId=10892570, @wolfs please take another look.

**Comment by lptr** [comment]
Another build: https://builds.gradle.org/viewQueued.html?itemId=10893758

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return true;
+    }
+
+    @Override
+    public Iterator<TaskStateChange> iterateContentChangesSince(FileCollectionSnapshot oldSnapshot, final String title, boolean includeAdded) {
+        return Iterators.transform(oldSnapshot.getContentSnapshots().entrySet().iterator(), new Function<Map.Entry<String, FileContentSnapshot>, TaskStateChange>() {
```
Comment (wolfs) [inline subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java]: Should we short-circuit here when the `oldSnapshot` is an `EmptyFileCollectionSnapshot`? Or do you think it doesn't make a difference?

File: subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshotTest.groovy [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+class EmptyFileCollectionSnapshotTest extends Specification {
+    def "comparing empty snapshot to regular snapshot shows entries added"() {
+        def snapshot = new DefaultFileCollectionSnapshot([
+            "file1.txt": new DefaultNormalizedFileSnapshot("file1.txt", new FileHashSnapshot(HashCode.fromInt(123))),
+            "file2.txt": new DefaultNormalizedFileSnapshot("file2.txt", new FileHashSnapshot(HashCode.fromInt(234))),
+        ], ORDERED, false)
+        expect:
+        snapshot.iterateContentChangesSince(EmptyFileCollectionSnapshot.INSTANCE, "test", false) as List == []
+        snapshot.iterateContentChangesSince(EmptyFileCollectionSnapshot.INSTANCE, "test", true) as List == [
+            FileChange.added("file1.txt", "test", FileType.Missing),
```
Comment (wolfs) [inline subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshotTest.groovy]: `FileType` should be `RegularFile`, right?

Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshotTest.groovy]: OK, so apparently `Iterator as List` returns something that equals everything... :o

Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshotTest.groovy]: Should be fixed now.

File: subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java (hunk 2 above)
Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/EmptyFileCollectionSnapshot.java]: I first added that, but I don't think it does.

**Comment by lptr** [comment]
More: https://builds.gradle.org/viewQueued.html?itemId=10894798

**Pull request merged**